services/object/src/api.js

```diff
@@ -33,4 +33,33 @@ builder.declare({
   return res.reply({});
 });
 
+builder.declare({
+  method: 'put',
+  route: '/download/:name',
```

Comment:
Maybe what is making me uncomfortable about object-service is that this is sorta pretending to be a download endpoint for the object when really it is just fetching metadata about how to download it. Even the title and description more accurately say what's going on here. What if we just called this `queryData`, `fetchData`, `objectMetadata`, etc? Thoughts?

Reply:
`:name` implies that the callers are picking the name of the download - is that true? I'd expect something like `:id` for a server-generated ID or a UUID.

I'm also uncomfortable, but I think that's because I don't have a clear idea of what the object service does in response to these requests. Is it just picking the "best" option from requested protocols? In that case, why would the best be any other than the first choice? Doesn't the client have to do some work to pick the request order? If the object server is just picking a protocol, then it seems we could reverse the logic, and the endpoint could return the menu of protocols, and the client could pick the one it likes.

But, I expect there is some work done to prepare a download via protocol. Some possible work:

* An authorized client could be asking for a public URL of a previously private download. The object server copies the file to a public bucket, or updates the metadata
* A client could ask for the gzipped version of a file. The object server uploads a gzipped version to a bucket and returns the URL
* A client asks for a URL to a private resource with baked-in credentials. The object server generates the URL with credentials in the query string
* A client asks for compressed content, with a list of compression formats. The server picks one that is ready to go, rather than re-compresses for the first requested format.

If this is the kind of work the object server does, then it makes more sense to me that the server picks. It could still be accomplished with a menu of protocol options, but it goes from two to three requests:

1. The client requests the metadata, maybe with a caching header for the last time the client asked. The server responds with the menu of protocol options, with data like "I can gzip this, but I haven't yet"
2. The client requests the selected protocol, with any protocol-specific options. The server does some work, and responds with the connection details, such as a URL
3. The client requests the download

But, that's probably a different object service. I'm happy to see where this one goes, and what use cases it supports.

Reply:
Who decides the name is a good question, that we haven't discussed (at least not recently).  I've been assuming that the client would choose the name, imagining names like `task/<taskId>/<runId>/<artifactName>`.  That allows pattern-matching on names in a useful in the `backendMap` configuration, for example putting `.*\.log$` in a reduced-redundancy bucket.  And it means that simple download URLs would "make sense" to the user, containing the taskId, runId, and artifact name that they expected.  It also helps with analysis and debugging.  And, it also leaves a bit of space for other (very speculative!) names like `worker-cache/<workerPoolId>/<cacheName>` or, for supporting mirroring, `replica/<replicaId>/<objectName>`.

Regarding the process for negotiation, for downloads I see this negotiation as a way to "decouple" download clients and servers, allowing us to introduce new functionality on the server with high confidence that the change won't break old clients because the old code (server-side and client-side) is maintained unchanged.  In that sense, "best" might just be "most recently implemented".  This can also support "weird" approaches -- the example we've been using is [bittorrent](https://docs.aws.amazon.com/AmazonS3/latest/dev/S3Torrent.html) -- that aren't supported by all backends.  

So hypothetically, imagine there's an `old-way` and `new-way`, as well as weird `bittorrent` and `blockchain` (it can do anything!) protocols.  The client makes a request for object `foo` with protocols `old-way`, `new-way`, `newer-way`, and `bittorrent`.  The object `foo` has an Azure backend, which doesn't support `bittorrent`, and the server doesn't know about `newer-way`, so the remaining possibilities are `old-way` and `new-way`.  The server selects `new-way` as it's the best of those two.  It calls the backend requesting protocol `new-way` and things proceed from there.

I think the key difference between the situations we've described is that you're seeing a lot more details involved in the negotiation (compression formats, cloud region, authorization, URL expiration time).  In my conception, those are parameters to the approach that are only consulted after the server chooses the (named) approach.

Reply:
> `:name` implies that the callers are picking the name of the download - is that true? I'd expect something like `:id` for a server-generated ID or a UUID.

Yes, in the current design, the caller chooses the name.

> I'm also uncomfortable, but I think that's because I don't have a clear idea of what the object service does in response to these requests. Is it just picking the "best" option from requested protocols? In that case, why would the best be any other than the first choice? Doesn't the client have to do some work to pick the request order? If the object server is just picking a protocol, then it seems we could reverse the logic, and the endpoint could return the menu of protocols, and the client could pick the one it likes.
> 
> But, I expect there is some work done to prepare a download via protocol. Some possible work:
> 
>     * An authorized client could be asking for a public URL of a previously private download. The object server copies the file to a public bucket, or updates the metadata
> 
>     * A client could ask for the gzipped version of a file. The object server uploads a gzipped version to a bucket and returns the URL
> 
>     * A client asks for a URL to a private resource with baked-in credentials. The object server generates the URL with credentials in the query string
> 
>     * A client asks for compressed content, with a list of compression formats. The server picks one that is ready to go, rather than re-compresses for the first requested format.
> 
> 
> If this is the kind of work the object server does, then it makes more sense to me that the server picks. It could still be accomplished with a menu of protocol options, but it goes from two to three requests:
> 
>     1. The client requests the metadata, maybe with a caching header for the last time the client asked. The server responds with the menu of protocol options, with data like "I can gzip this, but I haven't yet"
> 
>     2. The client requests the selected protocol, with any protocol-specific options. The server does some work, and responds with the connection details, such as a URL
> 
>     3. The client requests the download
> 
> 
> But, that's probably a different object service. I'm happy to see where this one goes, and what use cases it supports.

Regarding all of these ideas, I think they would be good enhancements for a future version of the object service.

Regarding compression negotiation and signed URLs, we could see it being added in future, as additional fields in the request body of this API method. Although compression could sometimes be negotiated in the download request directly, this may not always be possible, if the backing provider doesn't support it (e.g. we haven't found a way for S3 artifacts to be served supporting multiple content encodings) - so indeed the Object Service could handle this by creating additional S3 objects on-demand in this API call, if a new content encoding is required. Similarly, we could add an expiration property for signed URLs if the backing provider supports it.